# Segmenting and Clustering Neighborhoods in Toronto

## Part 1: Scrape the date from Wikipedia page

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from unicodedata import normalize

In [6]:
# scrape the data from the website using pd.read_html
table_TOR = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
len(table_Tor)

3

In [12]:
# there are 3 tables in scraped from the website and the table we want is the first one of the three table_TOR[0]
# assign the table to df and check the first and last 5 rows of the table
df = table_TOR[0]
print(df.head())
print(df.tail())

  Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront
    Postal Code       Borough  \
175         M5Z  Not assigned   
176         M6Z  Not assigned   
177         M7Z  Not assigned   
178         M8Z     Etobicoke   
179         M9Z  Not assigned   

                                         Neighbourhood  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  


In [11]:
# check the shpae of the table
df.shape

(180, 3)

In [34]:
#drop the row where borough is not assigned and and define the new dataframe as df1
df1 = df[df.Borough != 'Not assigned']
df1.reset_index(drop=True,inplace=True)

In [35]:
# Check if there are rows with Neighbourhood as 'Not assigned'
df1[df1.Neighbourhood =='Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

In [36]:
# Finally check how many rows we have
df1.shape

(103, 3)

In [39]:
df1['Postal Code']

0      M3A
1      M4A
2      M5A
3      M6A
4      M7A
5      M9A
6      M1B
7      M3B
8      M4B
9      M5B
10     M6B
11     M9B
12     M1C
13     M3C
14     M4C
15     M5C
16     M6C
17     M9C
18     M1E
19     M4E
20     M5E
21     M6E
22     M1G
23     M4G
24     M5G
25     M6G
26     M1H
27     M2H
28     M3H
29     M4H
      ... 
73     M4R
74     M5R
75     M6R
76     M7R
77     M9R
78     M1S
79     M4S
80     M5S
81     M6S
82     M1T
83     M4T
84     M5T
85     M1V
86     M4V
87     M5V
88     M8V
89     M9V
90     M1W
91     M4W
92     M5W
93     M8W
94     M9W
95     M1X
96     M4X
97     M5X
98     M8X
99     M4Y
100    M7Y
101    M8Y
102    M8Z
Name: Postal Code, Length: 103, dtype: object

### So far, we have scraped table from the wikipedia page and removed the rows with unassigned borough. The final pandas dataframe has 3 columns( PostalCode, Borough, and Neighborhood) and 103 rows.

## Part 2: Add coordinates to the dataframe for each postal code

In [41]:
# import the coordinates data from the link provided and assign the datafraom to coor
coor=pd.read_csv('https://cocl.us/Geospatial_data')

In [44]:
# chekc the column names of coor
coor.columns

Index(['Postal Code', 'Latitude', 'Longitude'], dtype='object')

In [53]:
# Merge the both dataframes df1 and coor to get the complete dataframe for next step
can=df1.merge(coor,on='Postal Code')

In [54]:
# Check the shape of tor and make sure we did not miss any data
can.shape

(103, 5)

In [55]:
# list the first five rows of the dataframe to check the strcuture
can.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

## Part 3 Explore and cluster the neighborhoods in Toronto

## Get the map of Toronto

In [66]:
# filter the dataframe so that we only work with only boroughs that contain the word 'Toronto'
tor = can[can['Borough'].str.contains("Toronto")]
tor.reset_index(drop=True,inplace=True)
tor.shape

(39, 5)

There is 39 rows in can which contain the word 'Toronto'.

In [71]:
import json # library to handle JSON files

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [75]:
# create map of Toronto using latitude and longitude values
latitude = 43.741667

longitude = -79.373333
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(tor['Latitude'], tor['Longitude'], tor['Borough'], tor['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## Download venue data from using Foursquare API

In [76]:
CLIENT_ID = 'H1LRIHQWKHBIXXYVKKZZY2YB3JJVSFNSHNAJOJODEX1F0Y1D' # your Foursquare ID
CLIENT_SECRET = 'Q2DHPFAQCKDU3TQN3G3DDXFVFJ5BFLHK3DL3CGBPJLJ1JZBF' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: H1LRIHQWKHBIXXYVKKZZY2YB3JJVSFNSHNAJOJODEX1F0Y1D
CLIENT_SECRET:Q2DHPFAQCKDU3TQN3G3DDXFVFJ5BFLHK3DL3CGBPJLJ1JZBF


In [77]:
tor.loc[0, 'Neighbourhood']

'Regent Park, Harbourfront'

In [78]:
neighborhood_latitude = tor.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = tor.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = tor.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


In [79]:
# type your answer here

LIMIT = 100 # limit of number of venues returned by Foursquare API



radius = 500 # define radius



# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=H1LRIHQWKHBIXXYVKKZZY2YB3JJVSFNSHNAJOJODEX1F0Y1D&client_secret=Q2DHPFAQCKDU3TQN3G3DDXFVFJ5BFLHK3DL3CGBPJLJ1JZBF&v=20180605&ll=43.6542599,-79.3606359&radius=500&limit=100'

In [80]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5fcd6e0be749390203f70445'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 44,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
 

In [81]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [82]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3     Body Blitz Spa East                  Spa  43.654735 -79.359874
4          Impact Kitchen           Restaurant  43.656369 -79.356980

In [83]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

44 venues were returned by Foursquare.


In [84]:
#Create a function to repeat the same process to all the neighborhoods in Toronto
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [85]:
tor_venues = getNearbyVenues(names=tor['Neighbourhood'],
                                   latitudes=tor['Latitude'],
                                   longitudes=tor['Longitude']
                                  )
tor_venues.shape

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

(1624, 7)

In [86]:
tor_venues.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Regent Park, Harbourfront                43.65426               -79.360636   
1  Regent Park, Harbourfront                43.65426               -79.360636   
2  Regent Park, Harbourfront                43.65426               -79.360636   
3  Regent Park, Harbourfront                43.65426               -79.360636   
4  Regent Park, Harbourfront                43.65426               -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4           Restaurant

In [87]:
tor_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Berczy Park                                                             55   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      68   
Christie                                                                16   
Church and Wellesley                                                    75   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              33   
Davisville North                                                         9   
Dufferin, Dovercourt Village                                            13   
First Canadian Place, Underground city                                 100   
Forest Hill North & West, Forest Hill Road Park                          4   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
High Park, The Junction South                                           25   
India Bazaar, The Beaches West                                          19   
Kensington Market, Chinatown, Grange Park                               74   
Lawrence Park                                                            3   
Little Portugal, Trinity                                                45   
Moore Park, Summerhill East                                              2   
North Toronto West, Lawrence Park                                       18   
Parkdale, Roncesvalles                                                  14   
Queen's Park, Ontario Provincial Government                             33   
Regent Park, Harbourfront                                               44   
Richmond, Adelaide, King                                               100   
Rosedale                                                                 4   
Roselawn                                                                 2   
Runnymede, Swansea                                                      33   
St. James Town                                                          85   
St. James Town, Cabbagetown                                             48   
Stn A PO Boxes                                                          96   
Studio District                                                         37   
Summerhill West, Rathnelly, South Hill, Forest ...                      14   
The Annex, North Midtown, Yorkville                                     19   
The Beaches                                                              4   
The Danforth West, Riverdale                                            43   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          34   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
Berczy Park                                                              55   
Brockton, Parkdale Village, Exhibition Place                             23   
Business reply mail Processing Centre, South Ce...                       16   
CN Tower, King and Spadina, Railway Lands, Harb...                       16   
Central Bay Street                                                       68   
Christie                                                                 16   
Church and Wellesley                                                     75   
Commerce Court, Victoria Hotel                                          100   
Davisville                                                     

In [88]:
print('There are {} uniques categories.'.format(len(tor_venues['Venue Category'].unique())))

There are 235 uniques categories.


## Analyze each neighbourhood

In [91]:
# one hot encoding
tor_onehot = pd.get_dummies(tor_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
tor_onehot['Neighbourhood'] = tor_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [tor_onehot.columns[-1]] + list(tor_onehot.columns[:-1])
tor_onehot = tor_onehot[fixed_columns]

tor_onehot.head()

Neighbourhood  Afghan Restaurant  Airport  Airport Food Court  \
0  Regent Park, Harbourfront                  0        0                   0   
1  Regent Park, Harbourfront                  0        0                   0   
2  Regent Park, Harbourfront                  0        0                   0   
3  Regent Park, Harbourfront                  0        0                   0   
4  Regent Park, Harbourfront                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Theater  Theme Restaurant  \
0                    0             0  ...        0                 0   
1                    0             0  ...        0                 0   
2                    0             0  ...        0                 0   
3                    0             0  ...        0                 0   
4                    0             0  ...        0                 0   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0                 0      0              0                              0   
1                 0      0              0                              0   
2                 0      0              0                              0   
3                 0      0              0                              0   
4                 0      0              0                              0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Yoga Studio  
0                 0                      0         0            0  
1                 0                      0         0            0  
2                 0                      0         0            0  
3                 0                      0         0            0  
4                 0                      0         0            0  

[5 rows x 236 columns]

In [92]:
tor_onehot.shape

(1624, 236)

In [93]:
# group rows by neighbourhood and by taking the mean of the frequency of occurrence of each category
tor_grouped = tor_onehot.groupby('Neighbourhood').mean().reset_index()
tor_grouped

Neighbourhood  Afghan Restaurant  \
0                                         Berczy Park           0.000000   
1        Brockton, Parkdale Village, Exhibition Place           0.000000   
2   Business reply mail Processing Centre, South C...           0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
4                                  Central Bay Street           0.000000   
5                                            Christie           0.000000   
6                                Church and Wellesley           0.013333   
7                      Commerce Court, Victoria Hotel           0.000000   
8                                          Davisville           0.000000   
9                                    Davisville North           0.000000   
10                       Dufferin, Dovercourt Village           0.000000   
11             First Canadian Place, Underground city           0.000000   
12    Forest Hill North & West, Forest Hill Road Park           0.000000   
13                           Garden District, Ryerson           0.000000   
14  Harbourfront East, Union Station, Toronto Islands           0.000000   
15                      High Park, The Junction South           0.000000   
16                     India Bazaar, The Beaches West           0.000000   
17          Kensington Market, Chinatown, Grange Park           0.000000   
18                                      Lawrence Park           0.000000   
19                           Little Portugal, Trinity           0.000000   
20                        Moore Park, Summerhill East           0.000000   
21                  North Toronto West, Lawrence Park           0.000000   
22                             Parkdale, Roncesvalles           0.000000   
23        Queen's Park, Ontario Provincial Government           0.000000   
24                          Regent Park, Harbourfront           0.000000   
25                           Richmond, Adelaide, King           0.000000   
26                                           Rosedale           0.000000   
27                                           Roselawn           0.000000   
28                                 Runnymede, Swansea           0.000000   
29                                     St. James Town           0.000000   
30                        St. James Town, Cabbagetown           0.000000   
31                                     Stn A PO Boxes           0.000000   
32                                    Studio District           0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...           0.000000   
34                The Annex, North Midtown, Yorkville           0.000000   
35                                        The Beaches           0.000000   
36                       The Danforth West, Riverdale           0.000000   
37           Toronto Dominion Centre, Design Exchange           0.000000   
38                     University of Toronto, Harbord           0.000000   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0    0.0000              0.0000        0.0000           0.000   
1    0.0000              0.0000        0.0000           0.000   
2    0.0000              0.0000        0.0000           0.000   
3    0.0625              0.0625        0.0625           0.125   
4    0.0000              0.0000        0.0000           0.000   
5    0.0000              0.0000        0.0000           0.000   
6    0.0000              0.0000        0.0000           0.000   
7    0.0000              0.0000        0.0000           0.000   
8    0.0000              0.0000        0.0000           0.000   
9    0.0000              0.0000        0.0000           0.000   
10   0.0000              0.0000        0.0000           0.000   
11   0.0000              0.0000        0.0000           0.000   
12   0.0000              0.0000        0.0000           0.000   
13   0.0000              0.0000        0.0000           0.000   
14   0.0000             

In [94]:
tor_grouped.shape

(39, 236)

In [96]:
# print the top 5 most common venues for each neighbourhood 
num_top_venues = 5

for hood in tor_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = tor_grouped[tor_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
            venue  freq
0     Coffee Shop  0.09
1     Cheese Shop  0.04
2      Restaurant  0.04
3          Bakery  0.04
4  Farmers Market  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.13
1       Nightclub  0.09
2     Coffee Shop  0.09
3  Breakfast Spot  0.09
4             Gym  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
           venue  freq
0           Park  0.06
1  Auto Workshop  0.06
2        Brewery  0.06
3     Skate Park  0.06
4  Burrito Place  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0    Airport Lounge  0.12
1   Airport Service  0.12
2             Plane  0.06
3   Harbor / Marina  0.06
4  Sculpture Garden  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.18
1                Café  0.06
2      Sandwich Pl

In [97]:
#define a function to sort the venues in descending order.
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [98]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = tor_grouped['Neighbourhood']

for ind in np.arange(tor_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(tor_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...            Skate Park   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery          Cocktail Bar              Beer Bar   
1        Breakfast Spot             Nightclub           Coffee Shop   
2           Pizza Place               Brewery         Burrito Place   
3       Airport Service              Boutique       Harbor / Marina   
4                  Café    Italian Restaurant        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Cheese Shop    Seafood Restaurant            Restaurant   
1          Climbing Gym         Burrito Place            Restaurant   
2            Restaurant        Farmers Market  Fast Food Restaurant   
3      Sculpture Garden         Boat or Ferry   Rental Car Location   
4   Japanese Restaurant       Bubble Tea Shop           Salad Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Farmers Market        Sandwich Place         Breakfast Spot  
1    Italian Restaurant          Intersection                    Bar  
2               Butcher      Recording Studio          Auto Workshop  
3                   Bar           Coffee Shop                  Plane  
4          Burger Joint      Department Store        Thai Restaurant

## Cluster Neighbourhoods using k means with k=5

In [101]:
# set number of clusters
kclusters = 5

tor_grouped_clustering = tor_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(tor_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:20] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1])

In [102]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

tor_merged = tor

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
tor_merged = tor_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

tor_merged.head() # check the last columns!

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               1           Coffee Shop   
1  43.662301 -79.389494               1           Coffee Shop   
2  43.657162 -79.378937               1        Clothing Store   
3  43.651494 -79.375418               1           Coffee Shop   
4  43.676357 -79.293031               4                   Pub   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery                  Park                   Pub   
1           Yoga Studio                 Diner            Restaurant   
2           Coffee Shop                  Café   Japanese Restaurant   
3                  Café          Cocktail Bar            Restaurant   
4                 Trail     Health Food Store          Neighborhood   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Café               Theater        Breakfast Spot   
1  Portuguese Restaurant                  Park           Music Venue   
2        Bubble Tea Shop        Cosmetics Shop                 Diner   
3               Beer Bar             Gastropub   American Restaurant   
4            Yoga Studio               Dog Run    Dim Sum Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Event Space            Shoe Store                  Hotel  
1    Mexican Restaurant    Italian Restaurant             Hobby Shop  
2        Lingerie Store      Ramen Restaurant     Italian Restaurant  
3        Farmers Market                 Hotel    Japanese Restaurant  
4                 Diner        Discount Store    Distribution Center

In [103]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(tor_merged['Latitude'], tor_merged['Longitude'], tor_merged['Neighbourhood'], tor_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters